# Cargamos librerias

In [1]:
# Importar librerías
import pandas as pd
from sqlalchemy import create_engine
import pymysql
from sqlalchemy import create_engine, text

# Creamos la conexión

In [2]:
import pandas as pd
from sqlalchemy import create_engine

# Parámetros de conexión (sin base de datos específica)
user = 'root'
password = 'password'
sqlServer = 'localhost'
port = 3306

# Conexión al servidor (sin base de datos)
connStr = f"mysql+mysqlconnector://{user}:{password}@{sqlServer}:{port}"
engine = create_engine(connStr)
conn = engine.connect()

print("✅ Conectado al servidor MySQL")

✅ Conectado al servidor MySQL


# Creamos la BBDD

In [ ]:
# Crear la base de datos
sSQL = "CREATE DATABASE IF NOT EXISTS TELCO"
conn.execute(sSQL)
print("✅ Base de datos 'TELCO' creada/verificada")

# Cargar los Datos desde el Dataset

In [3]:
telco = pd.read_csv(r"C:\Users\andy_\Documents\Datasets\Telco\Telco-Customer-Churn.csv")
telco

,customerID,gender,SeniorCitizen,Partner,Dependents,tenure,PhoneService,MultipleLines,InternetService,OnlineSecurity,...,DeviceProtection,TechSupport,StreamingTV,StreamingMovies,Contract,PaperlessBilling,PaymentMethod,MonthlyCharges,TotalCharges,Churn
0,7590-VHVEG,Female,0,Yes,No,1,No,No phone service,DSL,No,...,No,No,No,No,Month-to-month,Yes,Electronic check,29.85,29.85,No
1,5575-GNVDE,Male,0,No,No,34,Yes,No,DSL,Yes,...,Yes,No,No,No,One year,No,Mailed check,56.95,1889.5,No
2,3668-QPYBK,Male,0,No,No,2,Yes,No,DSL,Yes,...,No,No,No,No,Month-to-month,Yes,Mailed check,53.85,108.15,Yes
3,7795-CFOCW,Male,0,No,No,45,No,No phone service,DSL,Yes,...,Yes,Yes,No,No,One year,No,Bank transfer (automatic),42.30,1840.75,No
4,9237-HQITU,Female,0,No,No,2,Yes,No,Fiber optic,No,...,No,No,No,No,Month-to-month,Yes,Electronic check,70.70,151.65,Yes
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
7038,6840-RESVB,Male,0,Yes,Yes,24,Yes,Yes,DSL,Yes,...,Yes,Yes,Yes,Yes,One year,Yes,Mailed check,84.80,1990.5,No
7039,2234-XADUH,Female,0,Yes,Yes,72,Yes,Yes,Fiber optic,No,...,Yes,No,Yes,Yes,One year,Yes,Credit card (automatic),103.20,7362.9,No
7040,4801-JZAZL,Female,0,Yes,Yes,11,No,No phone service,DSL,Yes,...,No,No,No,No,Month-to-month,Yes,Electronic check,29.60,346.45,No
7041,8361-LTMKD,Male,1,Yes,No,4,Yes,Yes,Fiber optic,No,...,No,No,No,No,Month-to-month,Yes,Mailed check,74.40,306.6,Yes


# Limpiar Datos

In [6]:
telco['TotalCharges'] = pd.to_numeric(telco['TotalCharges'], errors='coerce').fillna(0)

# Conexión a MySQL y dar Estructura

In [12]:
# ============================================
# 1. Conexión a MySQL
# ============================================
from sqlalchemy import create_engine, text
engine_telco = create_engine("mysql+mysqlconnector://root:password@localhost:3306/TELCO")

# ============================================
# 2. Cargar a MySQL CON ESTRUCTURA DEFINIDA
# ============================================
from sqlalchemy import types

# 2.1 Definir el esquema de la tabla
schema = {
    'customerID': types.String(20),
    'gender': types.String(10),
    'SeniorCitizen': types.Integer,
    'Partner': types.String(5),
    'Dependents': types.String(5),
    'tenure': types.Integer,
    'PhoneService': types.String(5),
    'MultipleLines': types.String(20),
    'InternetService': types.String(20),
    'OnlineSecurity': types.String(20),
    'OnlineBackup': types.String(20),
    'DeviceProtection': types.String(20),
    'TechSupport': types.String(20),
    'StreamingTV': types.String(20),
    'StreamingMovies': types.String(20),
    'Contract': types.String(20),
    'PaperlessBilling': types.String(5),
    'PaymentMethod': types.String(30),
    'MonthlyCharges': types.Float,
    'TotalCharges': types.Float,
    'Churn': types.String(5)
}

# 2.2 Cargar los datos con el esquema definido
telco.to_sql('churn', engine, if_exists='replace', index=False, dtype=schema)
print("✅ Datos cargados con tipos definidos")

# 2.3 Añadir Primary Key
with engine.connect() as conn:
    conn.execute(text("ALTER TABLE churn ADD PRIMARY KEY (customerID)"))
    print("✅ Primary Key (customerID) añadida")

# ============================================
# 3. Verificar
# ============================================
with engine.connect() as conn:
    result = conn.execute(text("DESCRIBE churn"))
    print("\n📋 Estructura de la tabla:")
    for row in result:
        print(f"  • {row[0]} ({row[1]}) {'PK' if row[3] == 'PRI' else ''}")

✅ Datos cargados con tipos definidos
✅ Primary Key (customerID) añadida

📋 Estructura de la tabla:
  • customerID (b'varchar(20)') PK
  • gender (b'varchar(10)') 
  • SeniorCitizen (b'int') 
  • Partner (b'varchar(5)') 
  • Dependents (b'varchar(5)') 
  • tenure (b'int') 
  • PhoneService (b'varchar(5)') 
  • MultipleLines (b'varchar(20)') 
  • InternetService (b'varchar(20)') 
  • OnlineSecurity (b'varchar(20)') 
  • OnlineBackup (b'varchar(20)') 
  • DeviceProtection (b'varchar(20)') 
  • TechSupport (b'varchar(20)') 
  • StreamingTV (b'varchar(20)') 
  • StreamingMovies (b'varchar(20)') 
  • Contract (b'varchar(20)') 
  • PaperlessBilling (b'varchar(5)') 
  • PaymentMethod (b'varchar(30)') 
  • MonthlyCharges (b'float') 
  • TotalCharges (b'float') 
  • Churn (b'varchar(5)') 


## 1) Tipos de clientes

In [14]:
import pandas as pd

sSQL = """
SELECT 
    Contract,
    COUNT(*) AS total_clientes
FROM churn
GROUP BY Contract
"""

pd.read_sql(sSQL, con=engine_telco)

,Contract,total_clientes
0,One year,1473
1,Month-to-month,3875
2,Two year,1695


### 2) Clientes con multiples lineas, sin teléfono y solo con 1 linea

In [15]:
sSQL = """
SELECT 
    COUNT(DISTINCT customerID) AS total_clientes,
    MultipleLines
FROM churn
GROUP BY MultipleLines
"""

pd.read_sql(sSQL, con=engine_telco)

,total_clientes,MultipleLines
0,3390,No
1,682,No phone service
2,2971,Yes


### 3) ¿Cuantos clientes nos abandonan y cuantos continuan?

In [16]:
# Análisis de churn
sSQL = """
SELECT 
    Churn,
    COUNT(*) AS total,
    ROUND(100.0 * COUNT(*) / (SELECT COUNT(*) FROM churn), 2) AS porcentaje
FROM churn
GROUP BY Churn
"""
pd.read_sql(sSQL, con=engine_telco)

,Churn,total,porcentaje
0,No,5174,73.46
1,Yes,1869,26.54


### 4) ¿Cuales son los métodos de pagos más usados?

In [17]:
# Análisis de churn
sSQL = """
SELECT 
    PaymentMethod,
    COUNT(customerID) AS total_clientes,
    ROUND(100.0 * COUNT(*) / (SELECT COUNT(*) FROM churn), 2) AS porcentaje
FROM churn
GROUP BY PaymentMethod
"""
pd.read_sql(sSQL, con=engine_telco)

,PaymentMethod,total_clientes,porcentaje
0,Mailed check,1612,22.89
1,Electronic check,2365,33.58
2,Credit card (automatic),1522,21.61
3,Bank transfer (automatic),1544,21.92


### 5) ¿Los clientes con múltiples líneas tienen menos abandono?

In [18]:
# 2. ¿Los clientes con múltiples líneas tienen menos churn?
sSQL = """
SELECT 
    MultipleLines,
    COUNT(*) AS total,
    SUM(CASE WHEN Churn = 'Yes' THEN 1 ELSE 0 END) AS churn_yes,
    ROUND(100.0 * SUM(CASE WHEN Churn = 'Yes' THEN 1 ELSE 0 END) / COUNT(*), 2) AS tasa_churn
FROM churn
GROUP BY MultipleLines
ORDER BY tasa_churn
"""
pd.read_sql(sSQL, con=engine_telco)

,MultipleLines,total,churn_yes,tasa_churn
0,No phone service,682,170.0,24.93
1,No,3390,849.0,25.04
2,Yes,2971,850.0,28.61


### 6) Segmentación por tipos de cliente

In [19]:
sSQL = """
SELECT 
    CASE
        WHEN Tenure >= 24 THEN 'Loyal'
        WHEN Tenure BETWEEN 12 AND 23 THEN 'Regular'
        ELSE 'New'
    END AS CustomerType,
    COUNT(customerID) AS total_clientes,
    ROUND(100.0 * COUNT(*) / (SELECT COUNT(*) FROM churn), 2) AS porcentaje
FROM churn
GROUP BY CustomerType
ORDER BY total_clientes DESC

"""

pd.read_sql(sSQL, con=engine_telco)

,CustomerType,total_clientes,porcentaje
0,Loyal,3927,55.76
1,New,2069,29.38
2,Regular,1047,14.87


### 7) ¿Cúal es la tasa de abandono por tipo de cliente?

In [20]:
sSQL = """
SELECT 
    CASE
        WHEN Tenure >= 24 THEN 'Loyal'
        WHEN Tenure BETWEEN 12 AND 23 THEN 'Regular'
        ELSE 'New'
    END AS CustomerType,
    COUNT(*) AS total_clientes,
    SUM(CASE WHEN Churn = 'Yes' THEN 1 ELSE 0 END) AS churn_yes,
    ROUND(100.0 * SUM(CASE WHEN Churn = 'Yes' THEN 1 ELSE 0 END) / COUNT(*), 2) AS tasa_churn
FROM churn
GROUP BY CustomerType
ORDER BY total_clientes DESC
"""

pd.read_sql(sSQL, con=engine_telco)

,CustomerType,total_clientes,churn_yes,tasa_churn
0,Loyal,3927,561.0,14.29
1,New,2069,999.0,48.28
2,Regular,1047,309.0,29.51


De 100 clientes nuevos:
├── 48 se van en el primer año ❌
└── 52 se quedan

De esos 52 que se quedan:
├── 15 se van en el segundo año (29.51%) ❌
└── 37 llegan a ser leales ✅

### 8) Qué servicios tiene los NEW que nos abandonan?

In [21]:
sSQL = """
SELECT 
    InternetService,
    Contract,
    PaymentMethod,
    COUNT(*) AS clientes_perdidos
FROM churn
WHERE Churn = 'Yes' 
  AND Tenure <= 12
GROUP BY InternetService, Contract, PaymentMethod
ORDER BY clientes_perdidos DESC
LIMIT 10
"""

pd.read_sql(sSQL, con=engine_telco)

,InternetService,Contract,PaymentMethod,clientes_perdidos
0,Fiber optic,Month-to-month,Electronic check,449
1,DSL,Month-to-month,Electronic check,142
2,DSL,Month-to-month,Mailed check,102
3,Fiber optic,Month-to-month,Mailed check,77
4,Fiber optic,Month-to-month,Bank transfer (automatic),71
5,No,Month-to-month,Mailed check,63
6,Fiber optic,Month-to-month,Credit card (automatic),46
7,DSL,Month-to-month,Credit card (automatic),28
8,DSL,Month-to-month,Bank transfer (automatic),21
9,No,Month-to-month,Electronic check,11


Solo el combo "Fiber + Month-to-month + Electronic check":
449 clientes perdidos × cargo mensual promedio (~$70-80)
= ~$31,000 - $35,000 MENSUALES en ingresos perdidos
= ~$372,000 - $420,000 ANUALES 💸

### 9) ¿Cuanto dinero perdemos por cada combo?

In [22]:
sSQL = """
SELECT 
    InternetService,
    Contract,
    PaymentMethod,
    COUNT(*) AS clientes_perdidos,
    ROUND(AVG(MonthlyCharges), 2) AS cargo_promedio,
    ROUND(COUNT(*) * AVG(MonthlyCharges), 2) AS ingreso_perdido_mensual
FROM churn
WHERE Churn = 'Yes' 
  AND Contract = 'Month-to-month'
GROUP BY InternetService, Contract, PaymentMethod
ORDER BY ingreso_perdido_mensual DESC
LIMIT 5
"""

pd.read_sql(sSQL, con=engine_telco)

,InternetService,Contract,PaymentMethod,clientes_perdidos,cargo_promedio,ingreso_perdido_mensual
0,Fiber optic,Month-to-month,Electronic check,789,86.54,68281.50
1,Fiber optic,Month-to-month,Bank transfer (automatic),149,87.67,13062.10
2,Fiber optic,Month-to-month,Credit card (automatic),122,87.96,10731.15
3,DSL,Month-to-month,Electronic check,192,45.71,8776.45
4,Fiber optic,Month-to-month,Mailed check,102,82.42,8407.25


### 10 ¿Qué servicios adicionales tiene los clientes críticos?

In [23]:
sSQL = """
SELECT 
    OnlineSecurity,
    TechSupport,
    COUNT(*) AS total,
    ROUND(AVG(MonthlyCharges), 2) AS cargo_promedio
FROM churn
WHERE Churn = 'Yes'
  AND InternetService = 'Fiber optic'
  AND Contract = 'Month-to-month'
  AND PaymentMethod = 'Electronic check'
GROUP BY OnlineSecurity, TechSupport
ORDER BY total DESC
"""

pd.read_sql(sSQL, con=engine_telco)

,OnlineSecurity,TechSupport,total,cargo_promedio
0,No,No,651,85.18
1,No,Yes,65,93.84
2,Yes,No,65,90.88
3,Yes,Yes,8,102.99


### 11) ¿Cuales son los servicios contratados por tipo de cliente?

In [24]:
sSQL = """
SELECT 
    CustomerType,
    COUNT(*) AS total_clientes_en_segmento,
    
    -- PhoneService
    ROUND(100.0 * SUM(CASE WHEN PhoneService = 'Yes' THEN 1 ELSE 0 END) / COUNT(*), 2) AS pct_PhoneService_Yes,
    
    -- InternetService
    ROUND(100.0 * SUM(CASE WHEN InternetService = 'Fiber optic' THEN 1 ELSE 0 END) / COUNT(*), 2) AS pct_Fiber,
    ROUND(100.0 * SUM(CASE WHEN InternetService = 'DSL' THEN 1 ELSE 0 END) / COUNT(*), 2) AS pct_DSL,
    ROUND(100.0 * SUM(CASE WHEN InternetService = 'No' THEN 1 ELSE 0 END) / COUNT(*), 2) AS pct_No_Internet,
    
    -- StreamingTV
    ROUND(100.0 * SUM(CASE WHEN StreamingTV = 'Yes' THEN 1 ELSE 0 END) / COUNT(*), 2) AS pct_StreamingTV,
    
    -- StreamingMovies
    ROUND(100.0 * SUM(CASE WHEN StreamingMovies = 'Yes' THEN 1 ELSE 0 END) / COUNT(*), 2) AS pct_StreamingMovies

FROM (
    SELECT 
        CASE
            WHEN Tenure >= 24 THEN 'Loyal'
            WHEN Tenure BETWEEN 12 AND 23 THEN 'Regular'
            ELSE 'New'
        END AS CustomerType,
        PhoneService,
        InternetService,
        StreamingTV,
        StreamingMovies
    FROM churn
) AS subconsulta
GROUP BY CustomerType
ORDER BY CustomerType
"""

pd.read_sql(sSQL, con=engine_telco)

,CustomerType,total_clientes_en_segmento,pct_PhoneService_Yes,pct_Fiber,pct_DSL,pct_No_Internet,pct_StreamingTV,pct_StreamingMovies
0,Loyal,3927,90.27,44.82,34.91,20.27,48.71,49.40
1,New,2069,90.04,42.53,34.27,23.20,22.04,22.23
2,Regular,1047,91.02,43.55,32.57,23.88,32.28,31.71


Los clientes LOYAL tienen MÁS DEL DOBLE de contratación de servicios de streaming
que los clientes NEW (49% vs 22%)

### 12) ¿Los clientes con Streaming tienen menos abandono?

In [25]:
sSQL = """
SELECT 
    StreamingTV,
    StreamingMovies,
    COUNT(*) AS total,
    SUM(CASE WHEN Churn = 'Yes' THEN 1 ELSE 0 END) AS churn_yes,
    ROUND(100.0 * SUM(CASE WHEN Churn = 'Yes' THEN 1 ELSE 0 END) / COUNT(*), 2) AS tasa_churn
FROM churn
GROUP BY StreamingTV, StreamingMovies
ORDER BY tasa_churn
"""

pd.read_sql(sSQL, con=engine_telco)

,StreamingTV,StreamingMovies,total,churn_yes,tasa_churn
0,No internet service,No internet service,1526,113.0,7.40
1,Yes,Yes,1940,571.0,29.43
2,No,Yes,792,247.0,31.19
3,Yes,No,767,243.0,31.68
4,No,No,2018,695.0,34.44


### 13) ¿Clientes por género?

In [26]:
sSQL = """
SELECT 
    Gender,
    COUNT(customerID) AS total_clientes
FROM churn
GROUP BY Gender
ORDER BY total_clientes DESC
"""

pd.read_sql(sSQL, con=engine_telco)

,Gender,total_clientes
0,Male,3555
1,Female,3488


### 14) Tipos de clientes por género

In [27]:
sSQL = """
SELECT 
    Gender,
    CASE
        WHEN Tenure >= 24 THEN 'Loyal'
        WHEN Tenure BETWEEN 12 AND 23 THEN 'Regular'
        ELSE 'New'
    END AS customer_type,
    COUNT(customerID) AS total_clientes
FROM churn
GROUP BY Gender, customer_type
ORDER BY total_clientes DESC
"""

pd.read_sql(sSQL, con=engine_telco)

,Gender,customer_type,total_clientes
0,Male,Loyal,1993
1,Female,Loyal,1934
2,Male,New,1048
3,Female,New,1021
4,Female,Regular,533
5,Male,Regular,514


### 15) ¿Cuales son los servicios contratados por género?

In [28]:
sSQL = """
SELECT 
    Gender,
    -- Servicios de internet
    ROUND(100.0 * SUM(CASE WHEN InternetService = 'Fiber optic' THEN 1 ELSE 0 END) / COUNT(*), 2) AS pct_Fiber,
    ROUND(100.0 * SUM(CASE WHEN InternetService = 'DSL' THEN 1 ELSE 0 END) / COUNT(*), 2) AS pct_DSL,
    ROUND(100.0 * SUM(CASE WHEN InternetService = 'No' THEN 1 ELSE 0 END) / COUNT(*), 2) AS pct_No_Internet,
    
    -- Servicios de seguridad y soporte
    ROUND(100.0 * SUM(CASE WHEN OnlineSecurity = 'Yes' THEN 1 ELSE 0 END) / COUNT(*), 2) AS pct_OnlineSecurity,
    ROUND(100.0 * SUM(CASE WHEN TechSupport = 'Yes' THEN 1 ELSE 0 END) / COUNT(*), 2) AS pct_TechSupport,
    
    -- Servicios de streaming
    ROUND(100.0 * SUM(CASE WHEN StreamingTV = 'Yes' THEN 1 ELSE 0 END) / COUNT(*), 2) AS pct_StreamingTV,
    ROUND(100.0 * SUM(CASE WHEN StreamingMovies = 'Yes' THEN 1 ELSE 0 END) / COUNT(*), 2) AS pct_StreamingMovies,
    
    -- Servicios adicionales
    ROUND(100.0 * SUM(CASE WHEN OnlineBackup = 'Yes' THEN 1 ELSE 0 END) / COUNT(*), 2) AS pct_OnlineBackup,
    ROUND(100.0 * SUM(CASE WHEN DeviceProtection = 'Yes' THEN 1 ELSE 0 END) / COUNT(*), 2) AS pct_DeviceProtection
    
FROM churn
GROUP BY Gender
"""

pd.read_sql(sSQL, con=engine_telco)

,Gender,pct_Fiber,pct_DSL,pct_No_Internet,pct_OnlineSecurity,pct_TechSupport,pct_StreamingTV,pct_StreamingMovies,pct_OnlineBackup,pct_DeviceProtection
0,Female,44.52,34.06,21.42,29.44,29.44,38.85,39.31,35.15,34.49
1,Male,43.40,34.68,21.91,27.90,28.61,38.03,38.28,33.84,34.29


### 16) Tasa de abandono por ciudadanos mayores

In [29]:
sSQL = """
SELECT 
    SeniorCitizen,
    COUNT(customerID) AS total_clientes,
    ROUND(100.0 * SUM(CASE WHEN Churn = 'Yes' THEN 1 ELSE 0 END) / COUNT(*), 2) AS tasa_churn
FROM churn
GROUP BY SeniorCitizen
ORDER BY total_clientes DESC
"""

pd.read_sql(sSQL, con=engine_telco)

,SeniorCitizen,total_clientes,tasa_churn
0,0,5901,23.61
1,1,1142,41.68


### 17) ¿Qué servicios contratan los Senior que se van?

In [30]:
sSQL = """
SELECT 
    InternetService,
    Contract,
    PaymentMethod,
    COUNT(*) AS clientes_perdidos,
    ROUND(AVG(MonthlyCharges), 2) AS cargo_promedio
FROM churn
WHERE Churn = 'Yes' AND SeniorCitizen = 1
GROUP BY InternetService, Contract, PaymentMethod
ORDER BY clientes_perdidos DESC
LIMIT 10
"""

pd.read_sql(sSQL, con=engine_telco)

,InternetService,Contract,PaymentMethod,clientes_perdidos,cargo_promedio
0,Fiber optic,Month-to-month,Electronic check,260,87.56
1,DSL,Month-to-month,Electronic check,42,40.53
2,Fiber optic,Month-to-month,Credit card (automatic),42,88.65
3,Fiber optic,Month-to-month,Bank transfer (automatic),40,89.31
4,Fiber optic,Month-to-month,Mailed check,26,81.09
5,DSL,Month-to-month,Mailed check,13,40.28
6,Fiber optic,One year,Electronic check,11,106.02
7,Fiber optic,One year,Credit card (automatic),8,102.41
8,DSL,Month-to-month,Credit card (automatic),7,48.36
9,DSL,Month-to-month,Bank transfer (automatic),6,46.28


# Gráficos sobre las consultas:

In [31]:
# ============================================
# INFORME EJECUTIVO CON GRÁFICOS
# ============================================

import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns
from sqlalchemy import create_engine, text
import os
import warnings
warnings.filterwarnings('ignore')

# Configuración de estilo
plt.style.use('seaborn-v0_8-whitegrid')
sns.set_palette("husl")
plt.rcParams['figure.figsize'] = (10, 6)
plt.rcParams['font.size'] = 11

# ============================================
# 1. CONEXIÓN Y CARGA DE DATOS
# ============================================

# Conectar a MySQL
engine_telco = create_engine("mysql+mysqlconnector://root:password@localhost:3306/TELCO")

# Cargar datos
df = pd.read_sql("SELECT * FROM churn", con=engine_telco)

# Limpiar y preparar
df['TotalCharges'] = pd.to_numeric(df['TotalCharges'], errors='coerce').fillna(0)
df['Churn_Num'] = (df['Churn'] == 'Yes').astype(int)

# Crear rangos de antigüedad
df['Rango_Antiguedad'] = pd.cut(df['tenure'], 
                                 bins=[-1, 12, 24, 48, 200], 
                                 labels=['0-12 meses', '13-24 meses', '25-48 meses', '49+ meses'])

print("✅ Datos cargados correctamente")

# ============================================
# CREAR CARPETA PARA GRÁFICOS
# ============================================

os.makedirs('graficos_informe', exist_ok=True)
print("✅ Carpeta 'graficos_informe' creada")

# ============================================
# GRÁFICO 1: PROPORCIÓN DE ABANDONO (DONUT)
# ============================================

print("\n📊 Generando gráfico 1: Proporción de abandono...")

churn_counts = df['Churn'].value_counts()
fig, ax = plt.subplots(figsize=(8, 8))
wedges, texts, autotexts = ax.pie(
    churn_counts, 
    labels=['No Churn', 'Churn'],
    autopct='%1.1f%%', 
    colors=['#2A9D8F', '#E63946'], 
    startangle=90,
    textprops={'fontsize': 14, 'fontweight': 'bold'},
    explode=(0, 0.05),
    shadow=True
)

# Añadir círculo interior para efecto donut
centre_circle = plt.Circle((0, 0), 0.70, fc='white', linewidth=2, edgecolor='white')
fig.gca().add_artist(centre_circle)

# Añadir total en el centro
plt.text(0, 0, f"Total\n{len(df):,}", ha='center', va='center', 
         fontsize=18, fontweight='bold', color='#264653')

plt.title('Proporción de Abandono de Clientes', fontsize=16, fontweight='bold', pad=20)
plt.tight_layout()
plt.savefig('graficos_informe/01_proporcion_abandono.png', dpi=300, bbox_inches='tight')
plt.close()

# ============================================
# GRÁFICO 2: CHURN POR ANTIGÜEDAD
# ============================================

print("📊 Generando gráfico 2: Churn por antigüedad...")

churn_antiguedad = df.groupby('Rango_Antiguedad').apply(
    lambda x: (x['Churn'] == 'Yes').mean() * 100
).reset_index(name='Tasa_Churn')

# Ordenar correctamente
orden_antiguedad = ['0-12 meses', '13-24 meses', '25-48 meses', '49+ meses']
churn_antiguedad['Rango_Antiguedad'] = pd.Categorical(churn_antiguedad['Rango_Antiguedad'], 
                                                       categories=orden_antiguedad, ordered=True)
churn_antiguedad = churn_antiguedad.sort_values('Rango_Antiguedad')

fig, ax = plt.subplots(figsize=(10, 6))
colors = ['#E63946', '#F4A261', '#2A9D8F', '#264653']
bars = ax.bar(churn_antiguedad['Rango_Antiguedad'], churn_antiguedad['Tasa_Churn'], color=colors, edgecolor='white', linewidth=2)

ax.set_title('Tasa de Abandono por Antigüedad del Cliente', fontsize=16, fontweight='bold')
ax.set_xlabel('Meses como cliente', fontsize=12)
ax.set_ylabel('Tasa de Churn (%)', fontsize=12)

# Añadir etiquetas de valor en las barras
for bar, val in zip(bars, churn_antiguedad['Tasa_Churn']):
    ax.text(bar.get_x() + bar.get_width()/2, val + 0.5, f"{val:.1f}%", 
            ha='center', fontweight='bold', fontsize=12)

# Añadir línea de referencia (media general)
media_general = (df['Churn'] == 'Yes').mean() * 100
ax.axhline(y=media_general, color='gray', linestyle='--', alpha=0.7, 
           label=f'Media general: {media_general:.1f}%')
ax.legend()

plt.tight_layout()
plt.savefig('graficos_informe/02_churn_antiguedad.png', dpi=300, bbox_inches='tight')
plt.close()

# ============================================
# GRÁFICO 3: CHURN POR CONTRATO
# ============================================

print("📊 Generando gráfico 3: Churn por contrato...")

churn_contrato = df.groupby('Contract').apply(
    lambda x: (x['Churn'] == 'Yes').mean() * 100
).reset_index(name='Tasa_Churn')

fig, ax = plt.subplots(figsize=(8, 6))
colors = ['#E63946', '#F4A261', '#2A9D8F']
bars = ax.bar(churn_contrato['Contract'], churn_contrato['Tasa_Churn'], color=colors, edgecolor='white', linewidth=2)

ax.set_title('Tasa de Abandono por Tipo de Contrato', fontsize=16, fontweight='bold')
ax.set_xlabel('Tipo de Contrato', fontsize=12)
ax.set_ylabel('Tasa de Churn (%)', fontsize=12)

# Añadir etiquetas de valor
for bar, val in zip(bars, churn_contrato['Tasa_Churn']):
    ax.text(bar.get_x() + bar.get_width()/2, val + 0.5, f"{val:.1f}%", 
            ha='center', fontweight='bold', fontsize=12)

plt.tight_layout()
plt.savefig('graficos_informe/03_churn_contrato.png', dpi=300, bbox_inches='tight')
plt.close()

# ============================================
# GRÁFICO 4: CHURN POR MÉTODO DE PAGO
# ============================================

print("📊 Generando gráfico 4: Churn por método de pago...")

churn_pago = df.groupby('PaymentMethod').apply(
    lambda x: (x['Churn'] == 'Yes').mean() * 100
).sort_values(ascending=False).reset_index(name='Tasa_Churn')

fig, ax = plt.subplots(figsize=(10, 6))
colors = ['#E63946', '#F4A261', '#2A9D8F', '#264653']
bars = ax.bar(churn_pago['PaymentMethod'], churn_pago['Tasa_Churn'], color=colors, edgecolor='white', linewidth=2)

ax.set_title('Tasa de Abandono por Método de Pago', fontsize=16, fontweight='bold')
ax.set_xlabel('Método de Pago', fontsize=12)
ax.set_ylabel('Tasa de Churn (%)', fontsize=12)

for bar, val in zip(bars, churn_pago['Tasa_Churn']):
    ax.text(bar.get_x() + bar.get_width()/2, val + 0.5, f"{val:.1f}%", 
            ha='center', fontweight='bold', fontsize=11)

plt.xticks(rotation=15, ha='right')
plt.tight_layout()
plt.savefig('graficos_informe/04_churn_pago.png', dpi=300, bbox_inches='tight')
plt.close()

# ============================================
# GRÁFICO 5: IMPACTO DE SEGURIDAD Y SOPORTE
# ============================================

print("📊 Generando gráfico 5: Impacto de seguridad y soporte...")

# Crear combinación de seguridad
df['Perfil_Seguridad'] = df.apply(
    lambda row: 'Sin Seguridad' if row['OnlineSecurity'] == 'No' and row['TechSupport'] == 'No'
                else 'Solo Security' if row['OnlineSecurity'] == 'Yes' and row['TechSupport'] == 'No'
                else 'Solo Support' if row['OnlineSecurity'] == 'No' and row['TechSupport'] == 'Yes'
                else 'Security + Support',
    axis=1
)

# Ordenar perfiles
orden_perfiles = ['Sin Seguridad', 'Solo Security', 'Solo Support', 'Security + Support']
churn_seguridad = df.groupby('Perfil_Seguridad').apply(
    lambda x: (x['Churn'] == 'Yes').mean() * 100
).reset_index(name='Tasa_Churn')
churn_seguridad['Perfil_Seguridad'] = pd.Categorical(churn_seguridad['Perfil_Seguridad'], 
                                                       categories=orden_perfiles, ordered=True)
churn_seguridad = churn_seguridad.sort_values('Perfil_Seguridad')

fig, ax = plt.subplots(figsize=(10, 6))
colors = ['#E63946', '#F4A261', '#2A9D8F', '#264653']
bars = ax.bar(churn_seguridad['Perfil_Seguridad'], churn_seguridad['Tasa_Churn'], color=colors, edgecolor='white', linewidth=2)

ax.set_title('Impacto de la Seguridad y Soporte en el Abandono', fontsize=16, fontweight='bold')
ax.set_xlabel('Perfil de Seguridad / Soporte', fontsize=12)
ax.set_ylabel('Tasa de Churn (%)', fontsize=12)

for bar, val in zip(bars, churn_seguridad['Tasa_Churn']):
    ax.text(bar.get_x() + bar.get_width()/2, val + 0.5, f"{val:.1f}%", 
            ha='center', fontweight='bold', fontsize=11)

# Añadir anotación sobre reducción
reduccion = churn_seguridad[churn_seguridad['Perfil_Seguridad'] == 'Sin Seguridad']['Tasa_Churn'].values[0] - \
            churn_seguridad[churn_seguridad['Perfil_Seguridad'] == 'Security + Support']['Tasa_Churn'].values[0]
ax.annotate(f'⬇ Reducción del {reduccion:.1f}%', 
            xy=(3, churn_seguridad[churn_seguridad['Perfil_Seguridad'] == 'Security + Support']['Tasa_Churn'].values[0]),
            xytext=(2.5, 10),
            arrowprops=dict(arrowstyle='->', color='green', lw=2),
            fontsize=12, fontweight='bold', color='green')

plt.tight_layout()
plt.savefig('graficos_informe/05_impacto_seguridad.png', dpi=300, bbox_inches='tight')
plt.close()

# ============================================
# GRÁFICO 6: IMPACTO ECONÓMICO
# ============================================

print("📊 Generando gráfico 6: Impacto económico...")

# Datos del impacto económico
segmentos = ['Fiber+Month+Electronic', 'Senior Citizens', 'New+Streaming']
perdida_mensual = [68282, 35000, 15000]
colores_segmentos = ['#E63946', '#F4A261', '#2A9D8F']

fig, ax = plt.subplots(figsize=(10, 6))
bars = ax.bar(segmentos, perdida_mensual, color=colores_segmentos, edgecolor='white', linewidth=2)

ax.set_title('Ingreso Perdido Mensual por Segmento Crítico', fontsize=16, fontweight='bold')
ax.set_xlabel('Segmento de Clientes', fontsize=12)
ax.set_ylabel('Ingreso Perdido ($)', fontsize=12)

for bar, val in zip(bars, perdida_mensual):
    ax.text(bar.get_x() + bar.get_width()/2, val + 1000, f"${val:,.0f}", 
            ha='center', fontweight='bold', fontsize=12)

# Añadir total
total_perdida = sum(perdida_mensual)
ax.axhline(y=total_perdida, color='gray', linestyle='--', alpha=0.7, 
           label=f'Total: ${total_perdida:,.0f}/mes')
ax.legend()

plt.tight_layout()
plt.savefig('graficos_informe/06_impacto_economico.png', dpi=300, bbox_inches='tight')
plt.close()

# ============================================
# GRÁFICO 7: CORRELACIONES (Mapa de Calor)
# ============================================

print("📊 Generando gráfico 7: Mapa de correlaciones...")

# Preparar datos para correlación
df_corr = df[['tenure', 'MonthlyCharges', 'SeniorCitizen']].copy()
df_corr['Churn'] = (df['Churn'] == 'Yes').astype(int)
df_corr['Contract_Monthly'] = (df['Contract'] == 'Month-to-month').astype(int)
df_corr['Payment_Electronic'] = (df['PaymentMethod'] == 'Electronic check').astype(int)
df_corr['Internet_Fiber'] = (df['InternetService'] == 'Fiber optic').astype(int)
df_corr['OnlineSecurity'] = (df['OnlineSecurity'] == 'Yes').astype(int)

# Renombrar columnas para mejor visualización
df_corr.columns = ['Antigüedad', 'Cargo Mensual', 'Senior', 'Churn', 
                   'Contrato Mensual', 'Pago Electrónico', 'Fibra', 'Seguridad Online']

# Calcular correlaciones con Churn
correlaciones = df_corr.corr()[['Churn']].sort_values(by='Churn', ascending=False)

fig, ax = plt.subplots(figsize=(10, 8))
sns.heatmap(df_corr.corr(), annot=True, cmap='RdBu_r', center=0, fmt='.2f', 
            linewidths=1, annot_kws={'size': 10})

ax.set_title('Mapa de Correlaciones con Abandono (Churn)', fontsize=16, fontweight='bold')
plt.tight_layout()
plt.savefig('graficos_informe/07_correlaciones.png', dpi=300, bbox_inches='tight')
plt.close()

# ============================================
# GRÁFICO 8: EVOLUCIÓN DEL CHURN (Líneas)
# ============================================

print("📊 Generando gráfico 8: Evolución del churn...")

churn_por_tenure = df.groupby('tenure').apply(
    lambda x: (x['Churn'] == 'Yes').mean() * 100
).reset_index(name='Tasa_Churn')

fig, ax = plt.subplots(figsize=(12, 6))

# Línea principal
ax.plot(churn_por_tenure['tenure'], churn_por_tenure['Tasa_Churn'], 
        color='#E63946', linewidth=2.5, label='Tasa de Churn')

# Media general
media_general = (df['Churn'] == 'Yes').mean() * 100
ax.axhline(y=media_general, color='gray', linestyle='--', alpha=0.7, 
           label=f'Media general: {media_general:.1f}%')

# Sombreado de áreas
ax.fill_between(churn_por_tenure['tenure'], 0, churn_por_tenure['Tasa_Churn'], 
                alpha=0.2, color='#E63946')

# Añadir anotaciones
ax.axvline(x=12, color='#F4A261', linestyle=':', alpha=0.7, label='1 año')
ax.axvline(x=24, color='#2A9D8F', linestyle=':', alpha=0.7, label='2 años')
ax.text(12, 10, '1 año', ha='center', fontsize=10, color='#F4A261', fontweight='bold')
ax.text(24, 10, '2 años', ha='center', fontsize=10, color='#2A9D8F', fontweight='bold')

ax.set_title('Evolución de la Tasa de Churn por Meses de Antigüedad', fontsize=16, fontweight='bold')
ax.set_xlabel('Meses como cliente', fontsize=12)
ax.set_ylabel('Tasa de Churn (%)', fontsize=12)
ax.grid(True, alpha=0.3)
ax.legend()

plt.tight_layout()
plt.savefig('graficos_informe/08_evolucion_churn.png', dpi=300, bbox_inches='tight')
plt.close()

# ============================================
# GRÁFICO 9: BOXPLOT DE CARGOS MENSUALES
# ============================================

print("📊 Generando gráfico 9: Boxplot de cargos mensuales...")

fig, ax = plt.subplots(figsize=(10, 6))

# Crear boxplot
sns.boxplot(data=df, x='Churn', y='MonthlyCharges', 
            palette=['#2A9D8F', '#E63946'], width=0.5)

# Añadir puntos de datos (opcional)
sns.stripplot(data=df, x='Churn', y='MonthlyCharges', 
              alpha=0.3, color='gray', size=1.5)

ax.set_title('Distribución de Cargos Mensuales por Abandono', fontsize=16, fontweight='bold')
ax.set_xlabel('¿Abandonó?', fontsize=12)
ax.set_ylabel('Cargo Mensual ($)', fontsize=12)

# Añadir estadísticas
for i, churn_status in enumerate(['No', 'Yes']):
    valores = df[df['Churn'] == churn_status]['MonthlyCharges']
    ax.text(i, 0, f'Media: ${valores.mean():.0f}', ha='center', fontsize=10, fontweight='bold')

plt.tight_layout()
plt.savefig('graficos_informe/09_boxplot_cargos.png', dpi=300, bbox_inches='tight')
plt.close()

# ============================================
# GRÁFICO 10: TOP 5 DE RIESGO (Barras horizontales)
# ============================================

print("📊 Generando gráfico 10: Top factores de riesgo...")

factores = ['Fibra óptica (×9.15)', 'Contrato mensual (×4.30)', 
            'DSL (×3.25)', 'Contrato anual (×2.05)', 
            'Pago electrónico (×1.46)']
odds_ratios = [9.15, 4.30, 3.25, 2.05, 1.46]
colores_riesgo = ['#E63946', '#E63946', '#E63946', '#F4A261', '#F4A261']

fig, ax = plt.subplots(figsize=(10, 6))
bars = ax.barh(factores, odds_ratios, color=colores_riesgo, edgecolor='white', linewidth=1)

ax.set_title('Top 5 Factores que Más Aumentan el Riesgo de Abandono', fontsize=16, fontweight='bold')
ax.set_xlabel('Odds Ratio (Multiplicador de Riesgo)', fontsize=12)

# Añadir etiquetas
for bar, val in zip(bars, odds_ratios):
    ax.text(val + 0.1, bar.get_y() + bar.get_height()/2, f'×{val:.2f}', 
            va='center', fontweight='bold', fontsize=11)

# Añadir línea de referencia (1 = sin impacto)
ax.axvline(x=1, color='gray', linestyle='--', alpha=0.7, label='Sin impacto')
ax.legend()

plt.tight_layout()
plt.savefig('graficos_informe/10_top_riesgo.png', dpi=300, bbox_inches='tight')
plt.close()

# ============================================
# CREAR MARCAS DE AGUA EN GRÁFICOS
# ============================================

# Añadir texto "Telco Churn Analysis" en cada gráfico (opcional)
for file in os.listdir('graficos_informe'):
    if file.endswith('.png'):
        print(f"✅ Gráfico generado: {file}")

# ============================================
# RESUMEN FINAL
# ============================================

print("\n" + "="*60)
print("✅ TODOS LOS GRÁFICOS GENERADOS")
print("="*60)
print(f"\n📁 Carpeta: 'graficos_informe'")
print(f"📊 Total de gráficos: {len(os.listdir('graficos_informe'))}")
print("\nLista de gráficos generados:")
for i, file in enumerate(sorted(os.listdir('graficos_informe')), 1):
    print(f"  {i}. {file}")
print("\n" + "="*60)

✅ Datos cargados correctamente
✅ Carpeta 'graficos_informe' creada

📊 Generando gráfico 1: Proporción de abandono...
📊 Generando gráfico 2: Churn por antigüedad...
📊 Generando gráfico 3: Churn por contrato...
📊 Generando gráfico 4: Churn por método de pago...
📊 Generando gráfico 5: Impacto de seguridad y soporte...
📊 Generando gráfico 6: Impacto económico...
📊 Generando gráfico 7: Mapa de correlaciones...
📊 Generando gráfico 8: Evolución del churn...
📊 Generando gráfico 9: Boxplot de cargos mensuales...
📊 Generando gráfico 10: Top factores de riesgo...
✅ Gráfico generado: 01_proporcion_abandono.png
✅ Gráfico generado: 02_churn_antiguedad.png
✅ Gráfico generado: 03_churn_contrato.png
✅ Gráfico generado: 04_churn_pago.png
✅ Gráfico generado: 05_impacto_seguridad.png
✅ Gráfico generado: 06_impacto_economico.png
✅ Gráfico generado: 07_correlaciones.png
✅ Gráfico generado: 08_evolucion_churn.png
✅ Gráfico generado: 09_boxplot_cargos.png
✅ Gráfico generado: 10_top_riesgo.png
✅ Gráfico genera

# ML - Modelo de Regresión

## Paso 1: Preparamos los datos desde SQL

In [32]:
# Extraer los datos con variables dummy (factores convertidos a números)
sSQL = """
SELECT 
    Churn,
    tenure,
    MonthlyCharges,
    SeniorCitizen,
    CASE WHEN Contract = 'Month-to-month' THEN 1 ELSE 0 END AS Contract_Monthly,
    CASE WHEN Contract = 'One year' THEN 1 ELSE 0 END AS Contract_OneYear,
    CASE WHEN PaymentMethod = 'Electronic check' THEN 1 ELSE 0 END AS Payment_Electronic,
    CASE WHEN PaymentMethod = 'Mailed check' THEN 1 ELSE 0 END AS Payment_Mailed,
    CASE WHEN InternetService = 'Fiber optic' THEN 1 ELSE 0 END AS Internet_Fiber,
    CASE WHEN InternetService = 'DSL' THEN 1 ELSE 0 END AS Internet_DSL,
    CASE WHEN StreamingTV = 'Yes' AND StreamingMovies = 'Yes' THEN 1 ELSE 0 END AS Streaming_Both,
    CASE WHEN OnlineSecurity = 'Yes' THEN 1 ELSE 0 END AS OnlineSecurity,
    CASE WHEN TechSupport = 'Yes' THEN 1 ELSE 0 END AS TechSupport
FROM churn
"""

df_model = pd.read_sql(sSQL, con=engine_telco)
df_model['Churn'] = (df_model['Churn'] == 'Yes').astype(int)

print(df_model.head())

   Churn  tenure  MonthlyCharges  SeniorCitizen  Contract_Monthly  \
0      0       9            65.6              0                 0   
1      0       9            59.9              0                 1   
2      1       4            73.9              0                 1   
3      1      13            98.0              1                 1   
4      1       3            83.9              1                 1   

   Contract_OneYear  Payment_Electronic  Payment_Mailed  Internet_Fiber  \
0                 1                   0               1               0   
1                 0                   0               1               0   
2                 0                   1               0               1   
3                 0                   1               0               1   
4                 0                   0               1               1   

   Internet_DSL  Streaming_Both  OnlineSecurity  TechSupport  
0             1               0               0            1  
1       

## Paso 2: Regresión Logística

In [33]:
from sklearn.linear_model import LogisticRegression
from sklearn.model_selection import train_test_split
from sklearn.metrics import classification_report, roc_auc_score
import statsmodels.api as sm

# Variables predictoras
features = ['tenure', 'MonthlyCharges', 'SeniorCitizen', 
            'Contract_Monthly', 'Contract_OneYear',
            'Payment_Electronic', 'Payment_Mailed',
            'Internet_Fiber', 'Internet_DSL',
            'Streaming_Both', 'OnlineSecurity', 'TechSupport']

X = df_model[features]
y = df_model['Churn']

# Añadir constante para statsmodels
X_const = sm.add_constant(X)

# Modelo con statsmodels (te da p-valores e intervalos de confianza)
modelo = sm.Logit(y, X_const).fit()

print(modelo.summary())

Optimization terminated successfully.
         Current function value: 0.418924
         Iterations 8
                           Logit Regression Results                           
Dep. Variable:                  Churn   No. Observations:                 7043
Model:                          Logit   Df Residuals:                     7030
Method:                           MLE   Df Model:                           12
Date:                Thu, 18 Jun 2026   Pseudo R-squ.:                  0.2760
Time:                        15:07:34   Log-Likelihood:                -2950.5
converged:                       True   LL-Null:                       -4075.1
Covariance Type:            nonrobust   LLR p-value:                     0.000
                         coef    std err          z      P>|z|      [0.025      0.975]
--------------------------------------------------------------------------------------
const                 -2.7998      0.219    -12.769      0.000      -3.230      -2.370
tenur

## Paso 3: Interpretar los resultados

In [34]:
import numpy as np  

# Extraer coeficientes (el impacto de cada factor)
resultados = pd.DataFrame({
    'Variable': modelo.params.index,
    'Coeficiente': modelo.params.values,
    'P>|z|': modelo.pvalues,
    'Odds_Ratio': np.exp(modelo.params.values)
})

# Ordenar por impacto absoluto (excluyendo constante)
resultados = resultados[resultados['Variable'] != 'const']
resultados['Impacto_Absoluto'] = abs(resultados['Coeficiente'])
resultados = resultados.sort_values('Impacto_Absoluto', ascending=False)

print("\n🏆 FACTORES CON MÁS IMPACTO EN CHURN (ordenados):\n")
print(resultados[['Variable', 'Coeficiente', 'Odds_Ratio', 'P>|z|']].to_string())


🏆 FACTORES CON MÁS IMPACTO EN CHURN (ordenados):

                              Variable  Coeficiente  Odds_Ratio         P>|z|
Internet_Fiber          Internet_Fiber     2.213208    9.145008  9.393990e-16
Contract_Monthly      Contract_Monthly     1.457810    4.296541  4.238677e-17
Internet_DSL              Internet_DSL     1.178205    3.248536  3.081924e-13
Contract_OneYear      Contract_OneYear     0.715760    2.045741  4.232374e-05
Streaming_Both          Streaming_Both     0.648154    1.912008  5.924141e-10
Payment_Electronic  Payment_Electronic     0.377264    1.458290  8.533091e-07
OnlineSecurity          OnlineSecurity    -0.376747    0.686089  1.494627e-05
TechSupport                TechSupport    -0.322555    0.724296  2.826158e-04
SeniorCitizen            SeniorCitizen     0.288639    1.334610  4.793882e-04
tenure                          tenure    -0.031706    0.968791  8.414510e-49
Payment_Mailed          Payment_Mailed    -0.023726    0.976553  8.097653e-01
MonthlyCharge

# 📊 ANÁLISIS DE CHURN - TELECOMUNICACIONES
## Informe Ejecutivo Completo con Insights de EDA y Modelo de Regresión

---

## 1. DATOS GENERALES

| Métrica | Valor |
|---------|-------|
| Total de clientes analizados | 7,043 |
| Clientes que abandonaron (Churn = Yes) | 1,869 (26.5%) |
| Clientes activos (Churn = No) | 5,174 (73.5%) |

---

## 2. RESULTADOS DEL MODELO DE REGRESIÓN LOGÍSTICA

### 2.1 Factores que MÁS AUMENTAN el riesgo de churn (Odds Ratio > 1)

| # | Factor | Odds Ratio | Aumento de riesgo | Significado |
|---|--------|------------|-------------------|-------------|
| 1 | **Internet: Fiber optic** | 9.15 | **+815%** | El factor MÁS IMPORTANTE |
| 2 | **Contrato: Month-to-month** | 4.30 | **+330%** | Segundo más importante |
| 3 | **Internet: DSL** | 3.25 | **+225%** | Tercero |
| 4 | Contrato: One year | 2.05 | +105% | Sorpresa: también aumenta |
| 5 | Pago: Electronic check | 1.46 | +46% | Método de pago riesgoso |
| 6 | Senior Citizen | 1.33 | +33% | Edad avanzada |

### 2.2 Factores que PROTEGEN contra el churn (Odds Ratio < 1)

| # | Factor | Odds Ratio | Reducción de riesgo |
|---|--------|------------|---------------------|
| 1 | **Online Security** | 0.69 | **-31%** | 🛡️ El mejor protector |
| 2 | **Tech Support** | 0.72 | **-28%** | 🛡️ Segundo mejor |
| 3 | Tenure (cada mes adicional) | 0.97 | -3% | La antigüedad protege |

### 2.3 Factores SIN impacto significativo (p-valor > 0.05)

| Factor | Observación |
|--------|-------------|
| Payment_Mailed | No significativo |
| MonthlyCharges | No significativo |

---

## 3. HALLAZGOS CLAVE POR VARIABLE (EDA)

### 3.1 Antigüedad (CustomerType) - IMPACTO MUY ALTO 🔴

| Tipo de cliente | Antigüedad | Total | % del total | Tasa Churn |
|----------------|------------|-------|-------------|------------|
| New | < 12 meses | 2,069 | 29.38% | **48.28%** |
| Regular | 12-23 meses | 1,047 | 14.87% | **29.51%** |
| Loyal | ≥ 24 meses | 3,927 | 55.76% | **14.29%** |

**Conclusión:** Los clientes nuevos tienen **3.4 VECES MÁS** probabilidad de abandonar que los leales.

---

### 3.2 Senior Citizen (Edad) - IMPACTO MUY ALTO 🔴

| Senior Citizen | Total | Tasa Churn | Diferencia |
|----------------|-------|------------|------------|
| No (menor de 65 años) | 5,901 | 23.61% | base |
| **Sí (65 años o más)** | 1,142 | **41.68%** | **+18.07%** |

**Conclusión:** Los seniors tienen una tasa de abandono **76% MÁS ALTA** que el resto.

---

### 3.3 Contrato - IMPACTO MUY ALTO 🔴

El **100%** de los segmentos con mayor churn tienen contrato **Month-to-month**.
Los contratos anuales y de dos años tienen tasas de churn significativamente más bajas.

**Conclusión:** El contrato mes a mes es el segundo predictor más importante de abandono (Odds Ratio: 4.30).

---

### 3.4 Método de Pago - IMPACTO ALTO 🔴

| Método de pago | % de clientes | Odds Ratio | Posición en churn |
|----------------|---------------|------------|-------------------|
| Electronic check | 33.58% | 1.46 | El más frecuente en churn |
| Mailed check | 22.89% | 0.98 (no sig.) | Segundo más frecuente |
| Bank transfer (automatic) | 21.92% | - | Menor churn |
| Credit card (automatic) | 21.61% | - | Menor churn |

**Conclusión:** Los métodos de pago **NO automáticos** aumentan significativamente el riesgo.

---

### 3.5 Servicios de Streaming - HALLAZGO CRÍTICO (depende de la antigüedad) 🟠

| CustomerType | Streaming_Both | Tasa Churn | Interpretación |
|--------------|----------------|------------|----------------|
| **New** | Sí | **68.18%** | ⚠️ **MUY ALTO RIESGO** |
| Regular | Sí | 46.92% | 🟠 Riesgo alto |
| Loyal | Sí | 19.93% | 🟢 Bajo riesgo |

**Conclusión:** El streaming NO es malo en sí mismo. El problema es OFRECERLO A CLIENTES NUEVOS SIN COMPROMISO DE PERMANENCIA. Los clientes leales con streaming son los más valiosos.

---

### 3.6 Género - SIN IMPACTO 🟢

| Género | Tasa Churn | Diferencia |
|--------|------------|------------|
| Female | 26.92% | base |
| Male | 26.16% | -0.76% |

**Conclusión:** El género **NO** es un factor relevante. Ambos géneros se comportan igual.

---

### 3.7 Servicios que PROTEGEN (Online Security y Tech Support)

| Servicio | Odds Ratio | Reducción de riesgo |
|----------|------------|---------------------|
| Online Security | 0.69 | **-31%** |
| Tech Support | 0.72 | **-28%** |

**Conclusión:** Los clientes que contratan seguridad online o soporte técnico son significativamente más leales.

---

## 4. SEGMENTOS CRÍTICOS (ALTO RIESGO)

### 4.1 Segmento de ALTO RIESGO (General)

| Internet | Contrato | Método de pago | Clientes perdidos | Cargo promedio | Ingreso perdido MENSUAL |
|----------|----------|----------------|-------------------|----------------|------------------------|
| Fiber optic | Month-to-month | Electronic check | 789 | $86.54 | **$68,282** |
| Fiber optic | Month-to-month | Bank transfer | 149 | $87.67 | $13,062 |
| Fiber optic | Month-to-month | Credit card | 122 | $87.96 | $10,731 |
| DSL | Month-to-month | Electronic check | 192 | $45.71 | $8,776 |

**El segmento Fiber + Month-to-month + Electronic check pierde 5 VECES MÁS que el segundo.**

---

### 4.2 Segmento SENIOR de ALTO RIESGO

| Internet | Contrato | Método de pago | Clientes perdidos | % del total Senior perdido |
|----------|----------|----------------|-------------------|---------------------------|
| Fiber optic | Month-to-month | Electronic check | 260 | **54.6%** |
| Fiber optic | Month-to-month | Otros pagos | 108 | 22.7% |

**Conclusión:** El 54.6% de TODOS los seniors que se van tienen el perfil: **Fiber + Month-to-month + Electronic check**.

---

### 4.3 Segmento NEW con Streaming (ALERTA)

| CustomerType | Streaming_Both | Tasa Churn | Clientes afectados |
|--------------|----------------|------------|-------------------|
| New | Sí | **68.18%** | 264 |

**Conclusión:** Los clientes nuevos que contratan streaming tienen una tasa de abandono del 68%. Es un segmento de altísimo riesgo.

---

## 5. IMPACTO ECONÓMICO TOTAL

| Segmento | Clientes perdidos | Ingreso perdido MENSUAL | Ingreso perdido ANUAL |
|----------|-------------------|------------------------|----------------------|
| Fiber + Month-to-month + Electronic check (general) | 789 | $68,282 | **$819,384** |
| Senior Citizens (todos) | 476 | $35,000 | **$420,000** |
| New + Streaming_Both | 180 | $15,000 (estimado) | **$180,000** |
| **TOTAL ESTIMADO** | **1,445** | **$118,282** | **$1,419,384** |

---

## 6. RECOMENDACIONES ESTRATÉGICAS (basadas en regresión)

### 🔴 PRIORIDAD 1 (URGENTE) - Fiber + Month-to-month + Electronic check
**Odds Ratio: 9.15 × 4.30 × 1.46 = 57.4 (riesgo 57 VECES mayor)**
**Impacto:** 789 clientes | $68,282/mes

**Acciones:**
1. Ofrecer **descuento del 15-20%** por cambiar a pago automático
2. **Llamada personalizada** de retención a los 789 clientes
3. **Upgrade gratis de velocidad** por 3 meses
4. Ofrecer **migración a contrato anual** con 2 meses gratis
5. Incluir **Online Security y Tech Support gratis** 3 meses

---

### 🔴 PRIORIDAD 1B (URGENTE) - Senior Citizens con perfil crítico
**Odds Ratio Senior: 1.33**
**Impacto:** 260 clientes | $22,766/mes

**Acciones específicas:**
1. **Línea de soporte técnico prioritario** gratuita para seniors
2. **Descuento por edad** (15-20%) en planes de Fiber
3. **Plan Senior específico** con factura en papel y atención telefónica
4. **Llamada personalizada** a los 260 seniors identificados

---

### 🔴 PRIORIDAD 1C (URGENTE) - Clientes NEW con Streaming
**Tasa de churn: 68.18%**
**Impacto:** 264 clientes | $15,000/mes estimado

**Acciones específicas:**
1. **No ofrecer streaming sin permanencia mínima** (6-12 meses)
2. Vincular streaming a **contrato anual**
3. **Llamada de retención** a clientes new con streaming antes de fin de promoción
4. Ofrecer **descuento por continuidad** al renovar streaming

---

### 🟠 PRIORIDAD 2 - Promover servicios que PROTEGEN
**Online Security: -31% riesgo | Tech Support: -28% riesgo**

**Acciones:**
1. **Bundle obligatorio** Fiber + Online Security + Tech Support
2. **Descuento del 30%** por contratar ambos servicios
3. **Incluir gratis 3 meses** para clientes nuevos
4. Comunicar beneficio: "Clientes con seguridad online tienen 31% menos abandono"

---

### 🟠 PRIORIDAD 3 - Clientes NEW (primer año)
**Impacto:** 48.28% de churn en 2,069 clientes

**Acciones:**
1. **Programa de onboarding** (primeros 90 días)
2. **Descuento por permanencia** del 10% al cumplir el año
3. **Seguimiento mensual** los primeros 3 meses
4. Ofrecer **Online Security gratis** temporalmente

---

### 🟡 PRIORIDAD 4 - Migrar de Electronic check a pago automático
**Odds Ratio: 1.46 (+46% riesgo)**

**Acciones:**
1. **Descuento del 5-10%** por cambiar a pago automático
2. **Sorteos mensuales** entre clientes con pago automático
3. **Comunicación educativa** sobre beneficios

---

## 7. FACTORES SIN IMPACTO (NO INVERTIR)

| Factor | ¿Influye en churn? | Recomendación |
|--------|-------------------|---------------|
| **Género** | 🟢 NO | No segmentar campañas por género |
| **MonthlyCharges** | 🟢 NO | El precio solo no explica el churn |
| **Payment_Mailed** | 🟢 NO | No es prioritario |

---

## 8. MÉTRICAS DE ÉXITO

| Indicador | Objetivo | Plazo |
|-----------|----------|-------|
| Reducción de churn en segmento Fiber+Monthly+Electronic | -50% | 3 meses |
| Conversión a pago automático | +30% | 2 meses |
| Migración a contrato anual | +40% | 3 meses |
| Reducción de churn en Senior Citizens | -40% | 3 meses |
| Reducción de churn en New con streaming | -50% | 2 meses |
| Incremento de contratación de Online Security | +25% | 3 meses |
| ROI de campañas de retención | >300% | Trimestral |

---

## 9. PRÓXIMOS PASOS

1. ✅ Extraer lista de clientes del segmento crítico (789 clientes)
2. ✅ Extraer lista de seniors del segmento crítico (260 clientes)
3. ✅ Extraer lista de New con streaming (264 clientes)
4. 📞 Diseñar script de llamada para equipo de retención (3 perfiles distintos)
5. 💰 Calcular presupuesto para descuentos y upgrades
6. 🚀 Implementar campaña piloto con 100 clientes (del segmento crítico)
7. 📊 Medir resultados y ajustar estrategia
8. 🔄 Escalar campañas exitosas al resto de segmentos

---

## 10. CONCLUSIONES FINALES

### El perfil de cliente con MÁS probabilidad de abandonar (riesgo 57 VECES mayor):

> - 🔴 **Internet: Fiber optic** (×9.15)
> - 🔴 **Contrato: Month-to-month** (×4.30)
> - 🔴 **Pago: Electronic check** (×1.46)
> - 🔴 **Senior Citizen (65+ años)** (×1.33)
> - 🔴 **Nuevo (menos de 1 año)** (48.28% churn)
> - 🔴 **Con streaming** (68% churn si es nuevo)

### El perfil de cliente con MENOS probabilidad de abandonar (el ideal):

> - 🟢 **Contrato anual o de dos años**
> - 🟢 **Pago automático** (bank transfer o tarjeta)
> - 🟢 **Online Security + Tech Support** (-31% y -28% riesgo)
> - 🟢 **Leal (más de 2 años)** (14.29% churn)
> - 🟢 **Con streaming PERO con antigüedad** (20% churn)

---

## 11. FRASE EJECUTIVA PARA PRESENTACIÓN

> **"El 54.6% del churn total se concentra en clientes con Fiber optic, contrato month-to-month y pago con electronic check. Este segmento genera una pérdida anual estimada de $819,000. Convertir estos clientes a contrato anual y pago automático, junto con la venta cruzada de Online Security y Tech Support, podría reducir el churn hasta en un 50%."**

---

*Análisis realizado sobre dataset Telco-Customer-Churn*
*Herramientas: Python, SQL, MySQL, Pandas, Scikit-learn, Statsmodels*
*Fecha del análisis: [insertar fecha actual]*